Add support for torch.cond in vmap

guilhermeleobas opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #116051
* #116050
* #116159
* #116049
* __->__ #114523

Fixes: https://github.com/pytorch/pytorch/issues/114136

Patch enables conversion of a BatchedTensor into FakeTensor and write
torch.cond vmap support using torch.where

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

guilhermeleobas committed:
Add support for torch.cond in vmap

Fixes: https://github.com/pytorch/pytorch/issues/114136

Patch enables conversion of a BatchedTensor into FakeTensor and write
torch.cond vmap support using torch.where

[ghstack-poisoned]

guilhermeleobas committed:
Update on "[WIP] Add support for torch.cond in vmap"


Fixes: https://github.com/pytorch/pytorch/issues/114136

Patch enables conversion of a BatchedTensor into FakeTensor and write
torch.cond vmap support using torch.where

[ghstack-poisoned]

guilhermeleobas committed:
Update on "[WIP] Add support for torch.cond in vmap"


Fixes: https://github.com/pytorch/pytorch/issues/114136

Patch enables conversion of a BatchedTensor into FakeTensor and write
torch.cond vmap support using torch.where

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "[WIP] Add support for torch.cond in vmap"


Fixes: https://github.com/pytorch/pytorch/issues/114136

Patch enables conversion of a BatchedTensor into FakeTensor and write
torch.cond vmap support using torch.where

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "[WIP] Add support for torch.cond in vmap"


Fixes: https://github.com/pytorch/pytorch/issues/114136

Patch enables conversion of a BatchedTensor into FakeTensor and write
torch.cond vmap support using torch.where

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "[WIP] Add support for torch.cond in vmap"


Fixes: https://github.com/pytorch/pytorch/issues/114136

Patch enables conversion of a BatchedTensor into FakeTensor and write
torch.cond vmap support using torch.where

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "[WIP] Add support for torch.cond in vmap"


Fixes: https://github.com/pytorch/pytorch/issues/114136

Patch enables conversion of a BatchedTensor into FakeTensor and write
torch.cond vmap support using torch.where

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "[WIP] Add support for torch.cond in vmap"


Fixes: https://github.com/pytorch/pytorch/issues/114136

Patch enables conversion of a BatchedTensor into FakeTensor and write
torch.cond vmap support using torch.where

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas commented on torch/_subclasses/meta_utils.py:
```diff
@@ -305,6 +306,19 @@ def sym_sizes_strides_storage_offset(
                     if t.requires_grad and not is_leaf:
                         with torch.enable_grad():
                             r = r.clone()
+                elif is_batchedtensor(t):
+                    # Wraps a BatchedTensor in a FakeTensor
+                    sizes, strides, storage_offset = sym_sizes_strides_storage_offset(
+                        t, source
+                    )
+                    r = callback(
+                        lambda: torch.empty_strided(
+                            sizes,
+                            strides,
+                            dtype=t.dtype,
+                            device="meta",
+                        )
+                    )
```
@zou3519, I believe this part is correct. It transforms a `BatchedTensor` into a `FakeTensor`. Previously, you indicated that this portion should yield a 'BatchedTensor(FakeTensor(...)),' correct?"

zou3519 replied:
We should be converting a BatchedTensor into a BatchedTensor(FakeTensor(...)). Morally, we Fakeify Tensors so that Dynamo can run through the program on Tensors that do not have storage. The "Fake" versions of the tensors should have accurate metadata.

The problem with converting a BatchedTensor into a FakeTensor is that FakeTensor has different semantics from a BatchedTensor. For example, FakeTensor.sin() may return a tensor with certain sizes/strides, but BatchedTensor.sin() may return a Tensor with different strides (but the correct shapes). BatchedTensor.nonzero() will raise an error (because the operation is impossible), but FakeTensor.nonzero() will not.

guilhermeleobas committed:
Update on "Add support for torch.cond in vmap"


Fixes: https://github.com/pytorch/pytorch/issues/114136

Patch enables conversion of a BatchedTensor into FakeTensor and write
torch.cond vmap support using torch.where

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

zou3519 commented on torch/_subclasses/meta_utils.py:
```diff
@@ -556,10 +570,6 @@ def empty_create(inner_t, inner_src):
                                 # format here
                                 r = r.clone(memory_format=torch.preserve_format)
 
-                    # Graph-Break for wrapped tensors
-                    if torch._C._functorch.is_functorch_wrapped_tensor(t):
-                        return NotImplemented
```
we should still keep this here. BatchedTensor is not the only Functorch wrapped tensor. Instead we should change the condition to "if not is_batchedtnesor and is_functorch_wrapped_tensor"

zou3519 commented on torch/_higher_order_ops/cond.py:
```diff
@@ -404,3 +410,50 @@ def cond_func(ctx, pred, true_fn, false_fn, inputs):
+    else:
+        # predicate is known at this stage and it is a boolean expression or a
+        # tensor with one element.
+        true_fn = torch.vmap(true_fn, in_dims=in_dims)
+        false_fn = torch.vmap(false_fn, in_dims=in_dims)
```
Should probably happen in a follow-up PR, but for performance reasons, we should call something like resume_vmap here instead of torch.vmap (https://github.com/pytorch/pytorch/blob/main/torch/_functorch/vmap.py#L418).

The problem is: torch.vmap always returns a Tensor with the batch dimension, even if it didn't need it. In the following case:
```
torch.cond(pred, lambda x: torch.ones(3), lambda x: torch.zeros(3), (x,)
```

if we do
```py
x = torch.randn(2)
y = vmap(lambda x: torch.ones(3))(x)
```
We get that y is a tensor of shape (2, 3).  To make things more efficient, we want to return a Tensor with shape (3,).

guilhermeleobas committed:
Update on "Add support for torch.cond in vmap"


Fixes: https://github.com/pytorch/pytorch/issues/114136

Patch enables conversion of a BatchedTensor into FakeTensor and write
torch.cond vmap support using torch.where

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas commented on torch/_subclasses/meta_utils.py:
```diff
@@ -308,9 +311,9 @@ def sym_sizes_strides_storage_offset(
-                    sizes, strides, storage_offset = sym_sizes_strides_storage_offset(
-                        t, source
-                    )
+                    tensor = get_unwrapped(t)
+                    sizes = tensor.size()
+                    strides = tensor.stride()
```
It converts a `BatchedTensor(BatchedTensor(Tensor))` into a `BatchedTensor(FakeTensor)`. I'll make sure to update the code and include a test for nested BatchedTensors.

zou3519 replied:
(we should add a test for vmap(vmap(cond))(x))

guilhermeleobas replied:
Just a heads-up. Enabling the conversion of `BatchedTensor(Tensor)` into `BatchedTensor(FakeTensor)` allowed dynamo to compile code that previously would fallback to eager. Thus the number of failed tests in test_vmap

guilhermeleobas committed:
Update on "Add support for torch.cond in vmap"


Fixes: https://github.com/pytorch/pytorch/issues/114136

Patch enables conversion of a BatchedTensor into FakeTensor and write
torch.cond vmap support using torch.where

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Add support for torch.cond in vmap"


Fixes: https://github.com/pytorch/pytorch/issues/114136

Patch enables conversion of a BatchedTensor into FakeTensor and write
torch.cond vmap support using torch.where

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

zou3519 commented on test/functorch/test_control_flow.py:
```diff
@@ -1624,6 +1624,142 @@ def test_cond_tracing_with_valid_inputs(self, predType, innerFnType, nOperands,
+    @parametrize("nOperands", [1, 2])
+    @parametrize("nClosure", [0, 1])
+    @parametrize("nesting", [0])
+    def test_cond_vmap(self, predType, innerFnType, nOperands, nClosure, nesting):
```
Btw, before we merge this PR, we should fix our testing. We have a CI configuration that runs all of the functorch tests with Dynamo turned on, but for reasons that CI configuration doesn't actually run the functorch tests.

zou3519 replied:
I'll submit a couple of PRs for this

zou3519 replied:
Stack over at https://github.com/pytorch/pytorch/pull/115267

guilhermeleobas committed:
Update on "Add support for torch.cond in vmap"


Fixes: https://github.com/pytorch/pytorch/issues/114136

Patch enables conversion of a BatchedTensor into FakeTensor and write
torch.cond vmap support using torch.where

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Add support for torch.cond in vmap"


Fixes: https://github.com/pytorch/pytorch/issues/114136

Patch enables conversion of a BatchedTensor into FakeTensor and write
torch.cond vmap support using torch.where

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Add support for torch.cond in vmap"


Fixes: https://github.com/pytorch/pytorch/issues/114136

Patch enables conversion of a BatchedTensor into FakeTensor and write
torch.cond vmap support using torch.where

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas commented:
Will rebase this one after https://github.com/pytorch/pytorch/pull/115274 lands.

guilhermeleobas committed:
Update on "Add support for torch.cond in vmap"


Fixes: https://github.com/pytorch/pytorch/issues/114136

Patch enables conversion of a BatchedTensor into FakeTensor and write
torch.cond vmap support using torch.where

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Add support for torch.cond in vmap"


Fixes: https://github.com/pytorch/pytorch/issues/114136

Patch enables conversion of a BatchedTensor into FakeTensor and write
torch.cond vmap support using torch.where

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Add support for torch.cond in vmap"


Fixes: https://github.com/pytorch/pytorch/issues/114136

Patch enables conversion of a BatchedTensor into FakeTensor and write
torch.cond vmap support using torch.where

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/functorch/BatchedTensorImpl.cpp:
```diff
@@ -128,9 +128,19 @@ SymIntArrayRef BatchedTensorImpl::sym_strides_custom() const {
-  TORCH_CHECK(memory_format == MemoryFormat::Contiguous,
-      "NYI: querying is_contiguous inside of vmap for memory_format ",
-      "other than torch.contiguous_format");
+  // TORCH_CHECK is disable for now as there is some code in
+  // torch/_dynamo/tensor.py::TensorVariable::specialize that calls `is_contiguous_custom`
+  // with different memory formats. Since the only memory format supported on
+  // BatchedTensors is MemoryFormat::Contiguous, PyTorch just return false for
+  // the other ones.
+
```
Can you link the code? This sounds suspicious

guilhermeleobas replied:
Sure, here is the stacktrace if the TORCH_CHECK is enabled:
https://gist.github.com/guilhermeleobas/3a45544e201461a12eb9712ac889d1cb

zou3519 replied:
Okay, we shouldn't be changing BatchedTensor's is_contiguous semantics. Instead, somehow we want to pass is_contiguous=None to the TensorVariable when it gets built

guilhermeleobas replied:
Is this changing the `is_contiguous` semantics? Rather than raising an exception, this change returns `False` for any case that it is not `torch.contiguous_format`.

zou3519 replied:
That is changing the semantics, because instead of an error we end up doing something, right?

zou3519 commented on torch/_functorch/vmap.py:
```diff
@@ -274,10 +274,16 @@ def vmap_impl(func, in_dims, out_dims, randomness, chunk_size, *args, **kwargs):
-    )
+    from torch._dynamo import disable
+
+    @disable
```
Remember to remove the disable() when you get torch.compile to work with torch.vmap

zou3519 reviewed (commented):
The is_contiguous part is a bit suspicious to me, it also ends up changing the behavior in eager-mode PyTorch. Everything else looks good though.

guilhermeleobas committed:
Update on "Add support for torch.cond in vmap"


Fixes: https://github.com/pytorch/pytorch/issues/114136

Patch enables conversion of a BatchedTensor into FakeTensor and write
torch.cond vmap support using torch.where

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Add support for torch.cond in vmap"


Fixes: https://github.com/pytorch/pytorch/issues/114136

Patch enables conversion of a BatchedTensor into FakeTensor and write
torch.cond vmap support using torch.where

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Add support for torch.cond in vmap"


Fixes: https://github.com/pytorch/pytorch/issues/114136

Patch enables conversion of a BatchedTensor into FakeTensor and write
torch.cond vmap support using torch.where

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

lezcano commented on torch/_dynamo/variables/builder.py:
```diff
@@ -1451,7 +1451,10 @@ def _clone_input(value):
-        specialized_props = target_cls.specialize(example_value)
+        specialized_props = target_cls.specialize(
+            example_value,
+            test_is_contiguous=not torch._C._functorch.is_batchedtensor(example_value),
```
Agree this is a bit sus. What's going on here?

guilhermeleobas replied:
Calling `is_contiguous` on a BatchTensor with a memory format other than `contiguous_format` result in an exception. This code is just a switch to prevent the exception from happening. 

I also just pushed a commit that moves the check to inside `specialize` as suggested by Richard.

lezcano replied:
How did this work before this PR when compiling vmap? Was it just not tested?

zou3519 replied:
This combination (vmap over torch.cond) didn't work before this PR. To get it to work, we need to teach Dynamo about BatchedTensor (which we have not done before). BatchedTensor doesn't have semantics for is_contiguous in certain cases.

lezcano replied:
Oh, I thought this path would be hit already on regular `vmap`  (without `torch.cond`) already

zou3519 replied:
@lezcano in our previous vmap x torch.compile design Dynamo didn't peek into vmap (so it would never see any BatchedTensors). This ran us into subtle problems, so in our new design (that @guilhermeleobas is working on), Dynamo will actually see the BatchedTensors. Part of this PR (which says how to fake-ify a BatchedTensor) is necessary for that

lezcano replied:
thank you for the explanation!

guilhermeleobas committed:
Update on "Add support for torch.cond in vmap"


Fixes: https://github.com/pytorch/pytorch/issues/114136

Patch enables conversion of a BatchedTensor into FakeTensor and write
torch.cond vmap support using torch.where

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

zou3519 commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -158,13 +158,16 @@ def specialize(value: torch.Tensor):
-                ]
-            )
+            if torch._C._functorch.is_batchedtensor(value):
+                props["is_contiguous"] = ()
```
This should be props["is_contiguous"] = None, right? None causes Dynamo to actually query the value if it shows up. Otherwise it's as if this Tensor isn't contiguous for anything

zou3519 replied:
Actually, I'm kind of confused. The thing we want to happen is self.is_contiguous (for this TensorVariable) should be None, so that this branch gets triggered: https://github.com/pytorch/pytorch/blob/49d826bcd3de952eb84a33c89ed399a1a2821c15/torch/_dynamo/variables/tensor.py#L433-L448. What do we need to do to get there?

guilhermeleobas replied:
I guess this branch is only executed if one explicitly calls `tensor.is_contiguous()`:

```python
import torch

x = torch.ones(2, 3)

@torch.compile(backend="aot_eager", fullgraph=True)
def wrapper_fn(x):
    if x.is_contiguous():
        return x
    return x + 1

expected = wrapper_fn(x)
```

zou3519 replied:
Right, I guess my point is: under Dynamo, if x is a BatchedTensor, then x.is_contiguous(memory_format=channels_last) should raise an error. Same for all other memory_format that are not channels last. Can we add a test for this? It would probably look something like the following:
```
@torch.compile(backend="eager", fullgraph=True)
def f(x):
   if x.is_contiguous(memory_format=channels_last):
       return x.sin()
   return x.cos()

self.assertRaisesRegex(RuntimeError, "something"):
    vmap(f)(x)
```

zou3519 commented on torch/fx/passes/shape_prop.py:
```diff
@@ -35,11 +35,16 @@ def _extract_tensor_metadata(result : torch.Tensor) -> TensorMetadata:
-        torch.channels_last,
-        torch.channels_last_3d,
-    }
+    if torch._C._functorch.is_batchedtensor(result):
+        memory_formats = {
+            torch.contiguous_format,
+        }
```
How is `_extract_tensor_metadata` used in shape prop?

guilhermeleobas replied:
`extract_tensor_metadata` is invoked within the `CondHigherOrderVariable::call_function` to ensure the tensors returned by both the if and else branches have the same metadata. See:
https://github.com/pytorch/pytorch/blob/55ce4693ff2c0b6e50b8af323f36ecc7ff929638/torch/_dynamo/variables/higher_order_ops.py#L542-L560

zou3519 replied:
Okay this one's a bit trickier to resolve... cond needs to check that the result of true_fn and false_fn have the same properties. `_extract_tensor_metadata` is only meant to be used for actual concrete Tensors (and not Tensor subclasses). When we have vmap(torch.cond)(pred, true_fn, false_fn, x), what properties should we be checking? Should we check that true_fn(x) and false_fn(x) produces the same Tensor subclasses? (e.g. BatchedTensor vs regular Tensor)

cc @ydwu4

zou3519 replied:
I would be fine with us adding a `include_contiguity=True` flag to `_extract_tensor_metadata` and having the cond logic call extract_tensor_metadata(include_contiguity=False) to avoid this problem. Morally it seems like we only care about checking the metadata is the same when we're executing the cond with real non-subclass Tensors

guilhermeleobas committed:
Update on "Add support for torch.cond in vmap"


Fixes: https://github.com/pytorch/pytorch/issues/114136

Patch enables conversion of a BatchedTensor into FakeTensor and write
torch.cond vmap support using torch.where

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Add support for torch.cond in vmap"


Fixes: https://github.com/pytorch/pytorch/issues/114136

Patch enables conversion of a BatchedTensor into FakeTensor and write
torch.cond vmap support using torch.where

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Add support for torch.cond in vmap"


Fixes: https://github.com/pytorch/pytorch/issues/114136

Patch enables conversion of a BatchedTensor into FakeTensor and write
torch.cond vmap support using torch.where

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

zou3519 commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -158,13 +158,16 @@ def specialize(value: torch.Tensor):
-                    if value.is_contiguous(memory_format=x)
-                ]
-            )
+            if torch._C._functorch.is_batchedtensor(value):
```
Can you add a note saying that batchedtensor doesn't support all contiguity patterns, so we don't cache is_contiguous on it here?

guilhermeleobas committed:
Update on "Add support for torch.cond in vmap"


Fixes: https://github.com/pytorch/pytorch/issues/114136

Patch enables conversion of a BatchedTensor into FakeTensor and write
torch.cond vmap support using torch.where

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Add support for torch.cond in vmap"


Fixes: https://github.com/pytorch/pytorch/issues/114136

Patch enables conversion of a BatchedTensor into FakeTensor and write
torch.cond vmap support using torch.where

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Add support for torch.cond in vmap"


Fixes: https://github.com/pytorch/pytorch/issues/114136

Patch enables conversion of a BatchedTensor into FakeTensor and write
torch.cond vmap support using torch.where

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas commented:
@pytorchbot merge

guilhermeleobas committed:
Update on "Add support for torch.cond in vmap"


Fixes: https://github.com/pytorch/pytorch/issues/114136

Patch enables conversion of a BatchedTensor into FakeTensor and write
torch.cond vmap support using torch.where

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas committed:
Update on "Add support for torch.cond in vmap"


Fixes: https://github.com/pytorch/pytorch/issues/114136

Patch enables conversion of a BatchedTensor into FakeTensor and write
torch.cond vmap support using torch.where

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

guilhermeleobas commented:
@pytorchbot merge
